includes/abstracts/abstract-wp-job-manager-email-template.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+/**
+ * Abstract class for an email notification built using templates.
+ *
+ * @package wp-job-manager
+ *
+ * @since 1.31.0
+ */
+
+abstract class WP_Job_Manager_Email_Template extends WP_Job_Manager_Email {
+
+	/**
+	 * Get the rich text version of the email content.
+	 *
+	 * @return string
+	 */
+	public function get_rich_content() {
+		return $this->get_template( false );
+	}
+
+	/**
+	 * Get the plaintext version of the email content.
+	 *
+	 * @return string
+	 */
+	public function get_plain_content() {
+		if ( $this->has_template( true ) ) {
+			return $this->get_template( true );
+		}
+		return strip_tags( $this->get_rich_content() );
+	}
+
+	/**
+	 * Get the contents of a template.
+	 *
+	 * @param bool $plain_text
+	 * @return string|bool
+	 */
+	public function get_template( $plain_text = false ) {
+		$template = $this->locate_template( $plain_text );
+		if ( ! $template ) {
+			return false;
+		}
+		$args = $this->get_args();
+		$email = $this;
+
+		ob_start();
+		include $template;
+		return ob_get_clean();
+	}
+
+	/**
+	 * Check to see if a template exists for this email.
+	 *
+	 * @param bool $plain_text
+	 * @return bool
+	 */
+	public function has_template( $plain_text = false ) {
+		$template_file = $this->locate_template( $plain_text );
+		return $template_file && file_exists( $template_file );
+	}
+
+	/**
+	 * Locate template for this email.
+	 *
+	 * @param bool $plain_text
+	 * @return string
+	 */
+	protected function locate_template( $plain_text ) {
+		return locate_job_manager_template( $this->get_template_file_name( $plain_text ) );
+	}
+
+	/**
+	 * Generate the file name for the email template.
+	 *
+	 * @param bool $plain_text
+	 * @return string
+	 */
+	protected function get_template_file_name( $plain_text = false ) {
+		$class_name = get_class( $this );
+		// PHP 5.2: Using `call_user_func()` but `$class_name::get_key()` preferred.
+		$email_notification_key = call_user_func( array( $class_name, 'get_key') );
+		return WP_Job_Manager_Email_Notifications::get_template_file_name( $email_notification_key, $plain_text );
```

Comment:
This feels like a circular dependency to me. Since this file is responsible for rendering emails with a template, would it make sense for it to have knowledge of where to look for templates, rather than deferring to `WP_Job_Manager_Email_Notifications`? Not a show-stopper, but I'm uneasy with this extra dependency for some reason :)

Reply:
I had the logic for locating the template here, but actually moved it to `WP_Job_Manager_Email_Notifications` because of [upcoming work](https://github.com/Automattic/WP-Job-Manager/tree/add/job-details) where _segments_ of the emails will be rendered in that class. Otherwise, `WP_Job_Manager_Email_Notifications` would have to call a static on a incomplete abstract class (which I think is possible, but I feel uneasy with). 

The alternative was creating a global functions, but I try to keep those to a minimum.